dali/kernels/imgproc/convolution/laplacian_gpu.cuh

```diff
@@ -0,0 +1,338 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_H_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "dali/core/dev_buffer.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/convolution_gpu.h"
+#include "dali/kernels/imgproc/convolution/separable_convolution_gpu.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/scratch.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+
+namespace dali {
+namespace kernels {
+
+namespace laplacian {
+
+namespace detail {
+
+struct CastSampleDesc {
+  void* output;
+  const void* input;
+};
+
+template <typename Out, typename In>
+__global__ void BatchedCastKernel(const CastSampleDesc* samples,
+                                  const kernels::BlockDesc<1>* blocks) {
+  const auto& block = blocks[blockIdx.x];
+  const auto& sample = samples[block.sample_idx];
+  auto* out = reinterpret_cast<Out*>(sample.output);
+  const auto* in = reinterpret_cast<const In*>(sample.input);
+  for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+    out[x] = ConvertSat<Out>(in[x]);
+  }
+}
+
+}  // namespace detail
+
+/**
+ * @brief Computes convolution to obtain partial derivative in one of the dimensions.
+ * Convolution consits of `axes` windows, each to convolve along one dimension of the input data,
+ * where `deriv_axis`-th window is supposed to compute partial derivative along that axis,
+ * whereas the remaining windows should perform smoothing. Setup method expects
+ * a `has_smoothing` hint if all window sizes in non-derivative directions are one. If so,
+ * the smoothing convolutions can be skipped and only a single one-dimensional
+ * convolution in derivative direction is performed.
+ */
+template <typename Out, typename In, typename W, int axes, int deriv_axis, bool has_channels,
+          bool is_sequence>
+struct PartialDerivGpu {
+  using MultiDimConv = SeparableConvolutionGpu<Out, In, W, axes, has_channels, is_sequence>;
+  static constexpr int ndim = MultiDimConv::ndim;
+  static constexpr int sequence_axes = MultiDimConv::sequence_axes;
+  using SingleDimConv = ConvolutionGpu<Out, In, W, ndim, sequence_axes + deriv_axis, has_channels>;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const std::array<TensorListShape<1>, axes>& window_sizes,
+                           bool has_smoothing) {
+    has_smoothing_ = has_smoothing;
+    if (has_smoothing_) {
+      if (!multi_dim_impl_) {
+        single_dim_impl_.reset();
```

Comment:
I have my doubts about this - should we really free it? Is it possible that, due to varying parameters, `has_smoothing` changes from iteration to iteration? Perhaps `has_smoothing` should be a per-sample parameter and the batch split into smoothed and unsmoothed sub-batches? These are just wild thoughts.

Reply:
It is possible, and in that case it is reset and setup over again. I wasn't sure about that reset, if you think it's best not to have it, I can remove it.

I was thinking about that too. Actually, those kernels could be rewritten to reduce the number of kernel launches by increasing the memory consumption: on input one could in a one launch compute 

```
inter_left, inter_right = [dx(input), sx(input)]
```
then in the second launch

```
out_left, out_right = [sy(inter_left), dy(inter_right)]
```

And return `cast(out_left) + cast(out_right)`

And in that case it could be easier to exclude some samples from smoothing. But as the current approach seemed to performed reasonably when compared against gassuian_blur I decided to stick to it and apply the optimization with no smoothing only if the whole batch does not require it.

Reply:
Added the asserts